### PR TITLE
chainntnfs: extend scans for earlier height hints

### LINKED ISSUE #11182: [bug]: Force-close recovery can remain in waiting_close after restart despite sufficient confirmations
### Pre-Submission Checklist

- [x] I have searched the existing issues and believe this is a new bug.
- [x] I am not asking a question about how to use lnd, but reporting a bug (otherwise open a discussion).

### LND Version

v0.21.2-beta Commit: 29257ecc8892af0c882fa627b66a556190c84fd9

### LND Configuration

Relevant settings:
bitcoin.signet=true
bitcoin.node=bitcoind
db.backend=sqlite
db.use-native-sql=true

Private Taproot channels. No pending HTLCs in the controlled restart test. Peer connections use Tor.

### Backend Version

Bitcoin Core 29.3

### Backend Configuration

Public Signet, pruning enabled, no transaction index.

The closing block remained available locally. Bitcoin Core stayed running during the controlled LND stop/start.

### OS/Distribution

Debian 13 amd64, native systemd services.

### Bug Details & Steps to Reproduce

After a force close and graceful LND restart, the closing transaction had sufficient confirmations, but the channel remained in waiting_close_channels and its main local commitment output was not registered for sweeping.

Observed sequence on a two-node Signet setup:

1. Force-close a private Taproot channel.
2. Wait until LND recognizes the first confirmation at height H.
3. Gracefully stop LND before its three-confirmation notification is delivered. Leave Bitcoin Core running.
4. Let Core reach H+4, then start and unlock LND. Wait for synchronization.
5. After additional blocks, inspect lncli pendingchannels and lncli wallet pendingsweeps.

At H+6, the channel still appeared in waiting_close_channels with close_height=H and blocks_til_close_confirmed=0. Anchor sweep candidates were present, but the main local commitment output was absent.

A separate force close on the same node, without restarting during confirmation, advanced to pending_force_closing_channels after three confirmations and registered its main output for sweeping. That registration survived the subsequent restart.

The live failure may depend on startup registration order. The attached component test controls that order deterministically. We have not demonstrated permanent loss of funds.

Possibly related to #8120, #9532, and #10351. I could not establish whether these share the same underlying cause. Please redirect this report if the attached reproduction belongs under an existing issue.

### Expected Behavior

LND should recognize the historical confirmation after restart and advance the channel into force-close recovery once its confirmation threshold is met.

Sweeping may still wait for the output's normal maturity.

### Debug Information

The observed restart sequence, summarized without host or transaction identifiers:

- A six-confirmation subscription starts its lookup at H+4.
- Historical spend lookup finds the closing transaction at H.
- The three-confirmation watcher uses the later hint/search and misses that historical confirmation.

Suspected mechanism: TxNotifier.RegisterConf shares searches for the same transaction/script. A later-starting subscriber can register first, and an earlier hint from a subsequent subscriber does not trigger an expanded lookup. An empty lookup can also advance the cached hint.

The attached test adds one file to the pinned LND commit and changes no production code. Across five runs:

- No-restart control: PASS
- Three-confirmation subscriber first: PASS
- Six-confirmation subscriber first, initial scan completed: FAIL
- Six-confirmation subscriber first, initial scan pending: FAIL

The test intentionally exits nonzero on the affected commit. It uses synthetic data, models historical lookup, and uses LND's existing in-memory mock hint cache. It does not launch full daemons or use SQLite.

Separately, the same ordering failure was reproduced with actual Core historical lookup and a reopened SQLite height-hint cache. The correct hint survived reopening, and disabling cache queries did not resolve the ordering case. We have not established that this is SQLite-specific.

The attached ZIP contains the test-only patch, instructions, and local test output.

Could you confirm whether this matches a known notifier issue and advise on a supported recovery path for an already-stuck channel?

[lnd-confirmation-order-repro.zip](https://github.com/user-attachments/files/31975280/lnd-confirmation-order-repro.zip)

### Environment

Two Signet nodes with Tor peer connections. A node-management TUI submitted the force-close RPC; the controlled stop/start used systemctl.

The attached reproduction runs directly against LND's notifier code without the TUI or node-management code. Tested with Go 1.26.8.

### PR DESCRIPTION
**Pull request by Roasbeef**
In this PR, we make confirmation notifications independent of the order in which subscribers register height hints.

## Root Cause

The `TxNotifier` coalesces confirmation subscriptions for the same transaction and output script into a single historical scan. Before this change, the first registration selected that scan's lower bound. A later subscriber with an earlier height hint reused the first result without scanning the uncovered prefix.

This became reachable for channel closes after #10331 added a second close-transaction confirmation subscriber that waits for the capacity-scaled confirmation depth, including six confirmations for larger channels. After a restart, the wallet rebroadcaster can register its six-confirmation subscription at the current tip before the chain watcher restores its subscription with the transaction's actual height. If the close transaction was mined below the first hint, the shared scan missed it. Backends without an independent transaction index could not recover the confirmation.

Follow-up state-machine property testing found a second height-hint ownership issue. The final subscriber could cancel after its scan advanced the cached hint, then a reorg and restart could reuse that stale hint and skip a confirmation or spend on the replacement chain. Pending callbacks and Neutrino scan progress could also restore a hint after cancellation.

## Changes

* Track the earliest subscriber hint and the earliest dispatched scan height for each confirmation request.
* Dispatch only the uncovered historical prefix when a later subscriber supplies an earlier hint.
* Keep the shared height hint at the earliest outstanding range until all concurrent scans complete.
* Persist an accepted earlier range before dispatch so it survives another restart.
* Preserve bounded supplemental scan ranges in the Neutrino backend.
* Purge confirmation and spend hints after the final subscriber cancels.
* Retain pending scans and known in-memory results for later subscribers while keeping their durable hints purged.
* Keep subscriberless known results indexed until a reorg invalidates them.
* Exclude subscriberless requests from tip-driven hint updates.
* Serialize Neutrino spend-scan progress writes with cancellation.
* Preserve the first result for script-only spend requests so its reorg index remains consistent.
* Add registration-order, scan-completion-order, restart, cancellation, polling, pending-callback, and reorg regression coverage.
* Add a Rapid property that generates two to six subscribers and arbitrary scan-completion orders, then checks that their historical ranges form a gapless, non-overlapping partition and every subscriber receives the same result exactly once.
* Add a second Rapid state machine that varies three confirmation requests and three subscriber slots across registration, cancellation, blocks, reorgs, and notifier restarts. Its oracle checks exact updates, confirmations, negative confirmations, and duplicate delivery.

## Test Plan

* `go test ./chainntnfs -run 'TestConfirmation(RegistrationOrder|EqualHeightHintsShareScan)$' -count=20`
* `go test -race ./chainntnfs -run 'TestConfirmation(RegistrationOrder|EqualHeightHintsShareScan)$' -count=10`
* `go test ./chainntnfs -run '^TestConfirmationRegistrationOrderProperty$' -rapid.checks=1000 -count=1`
* `go test -race ./chainntnfs -run '^TestConfirmationRegistrationOrderProperty$' -rapid.checks=200 -count=1`
* `go test ./chainntnfs/... -count=1`
* `go test ./chainntnfs -run '^TestTxNotifierModelProperty$' -rapid.checks=10000 -count=1`
* `go test -race ./chainntnfs -run 'TestTxNotifier(Model|Stale|Canceled|Relevant|Repeated)' -rapid.checks=1000 -count=1`
* `go test -race ./chainntnfs/neutrinonotify -run '^TestCommitSpendHintIfActive$' -count=1`
* `make itest-only icase=open_psbt_channel_with_unstable_utxos backend=neutrino timeout=15m`
* `make itest-only icase=open_channel_with_unstable_utxos backend=neutrino timeout=15m`
* `make itest-only icase=channel_backup-restore_leased_zero_conf backend=neutrino timeout=15m`
* `make itest-only icase=channel_backup-restore_leased backend=neutrino timeout=15m`
* `make lint`
* `go vet ./chainntnfs/...`
* `go test ./contractcourt -run '^$'`

Fixes #11182

### CODE REVIEW

**Comment by Roasbeef** [comment]
Rebased. 

Pushed the property tests a bit harder and found some other lingering edge cases. The bulk of the diff is now extra tests.

File: chainntnfs/txnotifier.go
Comment (ViktorT-11) [inline chainntnfs/txnotifier.go:634]: It looks like the issue we're fixing here in this commit for "Confirmation subscribers", where a second subscriber registering with an earlier hint doesn't result in an actual scan, is also present for "Spend subscribers" below, i.e. via the `RegisterSpend` API endpoint IIUC.

I.e. a second "Spend subscriber" won't result in a scan if it's hint is earlier than the initial subscriber.
In theory, if future code for some reason introduces a second subscriber, or an external subscriber calls that endpoint with the current chain tip on a restart, this could lead to `lnd` missing a force-close after the restart if that second subscriber is registered first.

Does it make sense to fix that issue for "Spend subscribers" in this PR as well? Or should we tackle that in a follow-up PR?

File: chainntnfs/confirmation_registration_order_test.go [hunk 1]
```diff
@@ -0,0 +1,658 @@
...
+			testConfirmationRegistrationOrder(
+				t, testCase.completeBeforeEarly,
+				testCase.prefixFirst,
+				testCase.restartBeforePrefix,
+			)
+		})
+	}
+}
+
+func testConfirmationRegistrationOrder(t *testing.T, completeBeforeEarly,
```
Comment (ViktorT-11) [inline chainntnfs/confirmation_registration_order_test.go:53]: nit: This applies to quite a bit of the test functions in this PR, but IMO I think the docs here is quite lacking, and we'd improve readability if the docs were expanded. 

For example this function would benefit from some godocs despite being un-exported as it's quite a big function,  as well as from more comments within the function which explains what that part of the function is doing (that applies to quite many test functions within this PR).

File: chainntnfs/txnotifier.go [hunk 2]
```diff
@@ -728,6 +764,36 @@ func (n *TxNotifier) RegisterConf(txid *chainhash.Hash, pkScript []byte,
...
+		// retain the later cached hint and lose the earlier
+		// subscriber's range.
+		err := n.confirmHintCache.CommitConfirmHint(
+			startHeight, ntfn.ConfRequest,
+		)
+		if err != nil {
+			// The cache is an optimization, so a write failure
+			// does not prevent the live notifier from scanning the
+			// prefix.
+			Log.Debugf(
```
Comment (ViktorT-11) [inline chainntnfs/txnotifier.go:790]: nit: Should this really be a `debug` log level only? Perhaps a `warn` level is warranted, given that the message also details that this error doesn't prevent the live notifier from scanning the prefix.

**Review by ViktorT-11 — COMMENTED** [review]
Generally the fixes introduced by this PR looks good from my end :fire:! Adding a few feedback comments below, where especially the first comment is feedback which I think might make sense to address (perhaps in a follow-up PR).

File: chainntnfs/txnotifier.go [hunk 3]
```diff
@@ -845,24 +960,37 @@ func (n *TxNotifier) UpdateConfDetails(confRequest ConfRequest,
...
-	confSet.rescanStatus = rescanComplete
-
 	// The notifier has yet to reach the height at which the
 	// transaction/output script was included in a block, so we should defer
 	// until handling it then within ConnectTip.
 	if details == nil {
+		// Another scan is still checking an earlier range. An empty
+		// result cannot advance the shared height hint until all scans
+		// finish.
+		if confSet.pendingRescans > 0 {
```
Comment (ziggie1984) [inline chainntnfs/txnotifier.go:1171]: **Blocker (IMO):** a failed historical rescan permanently pins this counter above zero.

All three backends bail out of the dispatch goroutine on scan error without calling `UpdateConfDetails` (`bitcoindnotify/bitcoind.go:272`, `btcdnotify/btcd.go:387`, `neutrinonotify/neutrino.go:456` — the btcd one even carries a `TODO(wilmer): add retry logic if rescan fails?`). So `pendingRescans` never unwinds.

Before this PR that just meant "one scan stuck pending". Now it is considerably worse, because this guard and the one at L1015 make the counter authoritative for the whole set:

- the *other*, successful scan's result is swallowed here and `rescanStatus` never reaches `rescanComplete`;
- the persisted hint was already lowered to the earlier subscriber's value at registration (L789);
- `unconfirmedRequests()` skips non-`rescanComplete` sets, so nothing ever raises it again.

Net effect: a request that was healthy (`rescanComplete`, hint tracking tip) can be permanently downgraded to a frozen, very low persisted hint purely because a second subscriber registered with an earlier hint and that prefix scan happened to fail. Every subsequent restart then redoes the full rescan, forever.

The backends need to report dispatch *failure* back to the notifier (decrement the counter and restore the previous hint), not just log and return.

File: chainntnfs/txnotifier.go [hunk 4]
```diff
@@ -728,6 +770,36 @@ func (n *TxNotifier) RegisterConf(txid *chainhash.Hash, pkScript []byte,
...
+
+		Log.Debugf("Extending historical confirmation rescan for %v "+
+			"to include range %d-%d", ntfn.ConfRequest,
+			startHeight, endHeight)
+
+		// Persist the expanded obligation before dispatching it.
+		// Otherwise, a restart before this prefix completes would
+		// retain the later cached hint and lose the earlier
+		// subscriber's range.
+		err := n.confirmHintCache.CommitConfirmHint(
```
Comment (ziggie1984) [inline chainntnfs/txnotifier.go]: If this `CommitConfirmHint` fails we log and carry on, but `confSet.rescanStartHeight` is lowered unconditionally a few lines down (L840) while the *persisted* hint keeps the higher value.

The live notifier is fine, but the comment right above says persisting first exists precisely so "a restart before this prefix completes would retain the later cached hint and lose the earlier subscriber's range" — which is exactly what happens on this path. A write failure here silently degrades back to the bug this commit fixes.

Suggest either not lowering `rescanStartHeight` when the persist fails (so the set stays consistent with what's on disk), or treating it as a hard failure for the supplemental dispatch. Either way this is more than a debug log.

File: chainntnfs/txnotifier.go [hunk 5]
```diff
@@ -638,6 +658,9 @@ func (n *TxNotifier) RegisterConf(txid *chainhash.Hash, pkScript []byte,
 		return nil, err
 	}
 
+	n.Lock()
```
Comment (ziggie1984) [inline chainntnfs/txnotifier.go:661]: Moving the lock above the hint-cache query closes a real TOCTOU against `updateHints`, but it also pulls kvdb I/O inside the notifier lock, which is on the block-processing path (`ConnectTip`/`NotifyHeight`):

- `QueryConfirmHint` (read tx) now runs under the lock for **every** registration;
- the supplemental path at L789 does a `CommitConfirmHint` under it;
- `CancelConf`/`CancelSpend` now do `Purge{Confirm,Spend}Hint` under it, and those are `kvdb.Batch` → a bolt read-write transaction that can wait up to `MaxBatchDelay` before committing.

Two concerns:

1. **Latency** — every cancel is now a synchronous disk write serialized against block processing.
2. **Reentrancy** — bolt's write lock is not reentrant, so any caller that registers or cancels from inside an open `kvdb.Update` on the same DB would now self-deadlock. Previously `RegisterConf` never wrote the hint cache, so this hazard is new. I spot-checked the call sites in `contractcourt/*` and `peer/brontide.go:4947` and they look clean, but I didn't trace every closure — worth an explicit audit.

The purge in particular looks like it could happen outside the lock.

File: chainntnfs/txnotifier.go [hunk 6]
```diff
@@ -814,6 +889,46 @@ func (n *TxNotifier) CancelConf(confRequest ConfRequest, confID uint64) {
...
+
+	if len(confSet.ntfns) != 0 {
+		return
+	}
+
+	// Keep known details and pending scans alive for another subscriber.
+	// Their hint is still purged so a restart cannot trust progress that no
+	// active subscriber owns.
+	if confSet.details != nil || confSet.pendingRescans > 0 {
+		n.purgeConfirmHint(confRequest)
```
Comment (ziggie1984) [inline chainntnfs/txnotifier.go:901]: Purging the hint when the last subscriber goes away is the right call for correctness, but it's worth being explicit about the cost: after this, re-registering the same request rescans from the caller's original (possibly very old) hint rather than resuming.

For subsystems that cancel and re-register the same request during normal operation (contract resolvers, chain watcher), that could turn into repeated multi-thousand-block rescans. Would it be viable to keep the hint but mark it unowned — i.e. usable as a floor but not trusted as proof of scan progress — instead of deleting it outright?

Not blocking, but I think it deserves a sentence in the commit message either way.

File: chainntnfs/txnotifier.go [hunk 7]
```diff
@@ -715,6 +751,12 @@ func (n *TxNotifier) RegisterConf(txid *chainhash.Hash, pkScript []byte,
 	// another. When the rescan returns, this notification's details will be
 	// updated as well.
 	case rescanPending:
+		// The pending scan covers this subscriber's complete historical
+		// range. Its result will be shared with this notification.
+		if earlierHint && ntfn.HeightHint < confSet.rescanStartHeight {
```
Comment (ziggie1984) [inline chainntnfs/txnotifier.go:756]: Minor asymmetry with the `rescanComplete` break at L709, which additionally requires `confSet.details == nil` before taking the supplemental path.

I couldn't construct a reachable `rescanPending && details != nil` state (the tip handler sets both together, and L1026 sets `rescanComplete` alongside the details), so I don't think this is a live bug — but the two guards should probably read the same so a future change can't diverge them.

File: chainntnfs/txnotifier.go [hunk 8]
```diff
@@ -884,27 +1012,54 @@ func (n *TxNotifier) UpdateConfDetails(confRequest ConfRequest,
...
 
 	// Cache the details found in the rescan and attempt to dispatch any
 	// notifications that have not yet been delivered.
 	confSet.details = details
+	if len(confSet.ntfns) == 0 {
+		// Retain these details for a later subscriber. Track them for
+		// reorgs. There is no subscriber whose dispatch would add this
+		// index.
+		height := details.BlockHeight
+		reorgSafeHeight := height + n.reorgSafetyLimit
```
Comment (ziggie1984) [inline chainntnfs/txnotifier.go:1050]: When the confirmation is already deeper than `reorgSafetyLimit`, this skips the index insert, so the set is cached with details but never indexed — and `ConnectTip`'s maturity pruner only walks `confsByInitialHeight`. The subscriberless set then stays in `confNotifications` for the lifetime of the process.

master leaked these too (it never deleted sets on cancel at all), so this isn't a regression — but since this commit is specifically about aligning set lifetime with ownership, it seems worth closing here rather than leaving one path that still never gets collected.

File: chainntnfs/txnotifier.go [hunk 9]
```diff
@@ -1360,20 +1570,34 @@ func (n *TxNotifier) updateSpendDetails(spendRequest SpendRequest,
...
+		}
 	}
 
 	Log.Debugf("Updated spend hint to height=%v for confirmed spend "+
 		"request %v", details.SpendingHeight, spendRequest)
 
 	spendSet.details = details
+	if len(spendSet.ntfns) == 0 {
+		spendHeight := uint32(details.SpendingHeight)
+		txSet, exists := n.spendsByHeight[spendHeight]
```
Comment (ziggie1984) [inline chainntnfs/txnotifier.go:1591]: This index insert is missing the `reorgSafeHeight > n.currentHeight` guard that both of its siblings have — the conf-side equivalent at L1050 and `dispatchSpendDetails` at L1640.

Without it, a historical scan that lands a spend older than `reorgSafetyLimit` inserts into a `spendsByHeight` bucket whose maturity height has already passed. `ConnectTip` never sweeps it again and `DisconnectTip` only walks the height being disconnected, so both the `spendsByHeight` entry and the `spendNotifications` entry stay for the lifetime of the process.

Looks like a straightforward oversight — same guard as the other two should do it.

File: chainntnfs/neutrinonotify/neutrino.go [hunk 10]
```diff
@@ -828,6 +828,20 @@ func (n *NeutrinoNotifier) RegisterSpendNtfn(outpoint *wire.OutPoint,
 		return ntfn.Event, nil
 	}
 
+	// Mark the historical scan as canceled before removing its subscriber
+	// from the TxNotifier. This lets progress writes observe the same
+	// ownership boundary as the hint purge performed by cancellation.
+	var scanMtx sync.Mutex
+	var scanCanceled bool
+	originalCancel := ntfn.Event.Cancel
```
Comment (ziggie1984) [inline chainntnfs/neutrinonotify/neutrino.go]: `ntfn.Event.Cancel` is reassigned here *after* `RegisterSpend` has already inserted the event into the `TxNotifier`'s internal map, and without holding the notifier lock. That's an unsynchronized write to a field of an object that is already reachable from shared state.

It's benign today — I grepped and nothing inside `chainntnfs` calls `Event.Cancel`, which is why `-race` stays quiet — but it's a race the moment anything on the notifier side ever does, and that's a non-obvious invariant for a future change to preserve.

Cheapest fix is to build the wrapper before calling `RegisterSpend`, or to carry the cancellation flag on the registration struct instead of patching the published event.

File: chainntnfs/neutrinonotify/neutrino.go [hunk 11]
```diff
@@ -828,6 +828,20 @@ func (n *NeutrinoNotifier) RegisterSpendNtfn(outpoint *wire.OutPoint,
...
+	ntfn.Event.Cancel = func() {
+		scanMtx.Lock()
+		defer scanMtx.Unlock()
+
+		scanCanceled = true
```
Comment (ziggie1984) [inline chainntnfs/neutrinonotify/neutrino.go:841]: This holds `scanMtx` across `originalCancel()`, which goes `CancelSpend` → `n.Lock()` → `PurgeSpendHint` (a bolt RW transaction).

There's a pre-existing hazard where `ConnectTip` can block holding `n.Lock()` on a client that isn't draining its `Event.Spend` channel. If that same client is the one calling `Cancel()`, it now blocks holding `scanMtx`, which in turn wedges the neutrino historical-rescan goroutine in `commitSpendHintIfActive`. The client/`ConnectTip` interaction isn't new, but dragging the rescan goroutine into it is.

An atomic bool (or a try-lock on the progress write) would give the same ownership boundary without holding a mutex across a lock acquisition and a disk write.

File: chainntnfs/neutrinonotify/neutrino.go [hunk 12]
```diff
@@ -994,7 +1028,13 @@ func (n *NeutrinoNotifier) RegisterConfirmationsNtfn(txid *chainhash.Hash,
 	currentHeight := uint32(n.bestBlock.Height)
 	n.bestBlockMtx.RUnlock()
 
-	ntfn.HistoricalDispatch.EndHeight = currentHeight
+	// An initial scan ends at the txNotifier height observed during
+	// registration. Extend that scan through any blocks connected while the
+	// filter update was in flight. A supplemental scan ends below that
+	// height and must retain its non-overlapping prefix boundary.
+	if ntfn.HistoricalDispatch.EndHeight == ntfn.Height {
```
Comment (ziggie1984) [inline chainntnfs/neutrinonotify/neutrino.go:1035]: This infers "initial scan" from an equality that can hold accidentally rather than from anything the `TxNotifier` actually states.

If a set previously took the `startHeight > currentHeight` path, `rescanStartHeight` is set to `currentHeight + 1`, so a later supplemental dispatch gets `EndHeight == currentHeight == ntfn.Height` and is treated as initial (extended to the neutrino tip). It happens to be harmless in that specific case — nothing had been scanned, so extending is correct — but it's an implicit coupling between two files that is easy to break.

An explicit `Supplemental bool` (or similar) on `HistoricalConfDispatch` would make the contract legible from both sides.

**Review by ziggie1984 — CHANGES_REQUESTED** [review]
important bug fix

File: chainntnfs/txnotifier.go
Comment (Roasbeef) [reply chainntnfs/txnotifier.go:634]: Yeah my goal is to really address all related notification related bug fixes in this PR. Expands the scope a bit, but most of the PR is tests.

File: chainntnfs/confirmation_registration_order_test.go (hunk 1 above)
Comment (Roasbeef) [reply chainntnfs/confirmation_registration_order_test.go:53]: Will add additional godoc and normal comments.

File: chainntnfs/txnotifier.go (hunk 2 above)
Comment (Roasbeef) [reply chainntnfs/txnotifier.go:790]: warm sgtm

File: chainntnfs/txnotifier.go (hunk 3 above)
Comment (Roasbeef) [reply chainntnfs/txnotifier.go:988]: > he btcd one even carries a TODO(wilmer): add retry logic if rescan fails?). So pendingRescans never unwinds.

Will look into this

File: chainntnfs/txnotifier.go (hunk 4 above)
Comment (Roasbeef) [reply chainntnfs/txnotifier.go]: Hmm, so if the write fails here and we move forward, that's fine as then we'll start the rescan from that height from there on. 

Will look into the comment or log, and also the suggestions here.

File: chainntnfs/txnotifier.go (hunk 5 above)
Comment (Roasbeef) [reply chainntnfs/txnotifier.go:754]: > Reentrancy — bolt's write lock is not reentrant, so any caller that registers or cancels from inside an open kvdb.Update on the same DB would now self-deadlock. Previously RegisterConf never wrote the hint cache, so this hazard is new. I spot-checked the call sites in contractcourt/* and peer/brontide.go:4947 and they look clean, but I didn't trace every closure — worth an explicit audit.


Will pull on this thread to find any lingering issues there, will do a clean audit of that area

File: chainntnfs/txnotifier.go (hunk 6 above)
Comment (Roasbeef) [reply chainntnfs/txnotifier.go:901]: Hmm yeah good point. Part of the issue was the cache being sort of poisoned in the face of re-orgs. We could consider storing a different cached version that can instantly respond with a safe depth? Will look into that option.

File: chainntnfs/txnotifier.go (hunk 8 above)
Comment (Roasbeef) [reply chainntnfs/txnotifier.go:1050]: Will fix this while we're here
